contracts/src/v0.8/upkeeps/ERC20BalanceMonitor.sol

```diff
@@ -0,0 +1,289 @@
+    if (addresses.length != minBalances.length || addresses.length != topUpAmounts.length) {
+      revert InvalidWatchList();
+    }
+    address[] memory oldWatchList = s_watchList;
+    for (uint256 idx = 0; idx < oldWatchList.length; idx++) {
+      s_targets[oldWatchList[idx]].isActive = false;
+    }
+    for (uint256 idx = 0; idx < addresses.length; idx++) {
+      if (s_targets[addresses[idx]].isActive) {
+        revert DuplicateAddress(addresses[idx]);
+      }
+      if (addresses[idx] == address(0)) {
+        revert InvalidWatchList();
+      }
+      if (topUpAmounts[idx] == 0) {
+        revert InvalidWatchList();
+      }
+      s_targets[addresses[idx]] = Target({
+        isActive: true,
+        minBalance: minBalances[idx],
+        topUpAmount: topUpAmounts[idx],
+        lastTopUpTimestamp: 0
+      });
+    }
+    s_watchList = addresses;
+    emit WatchlistUpdated(oldWatchList, addresses);
+  }
+
+  /**
+   * @notice Gets a list of subscriptions that are underfunded.
+   * @return list of subscriptions that are underfunded
+   */
+  function getUnderfundedAddresses() public view returns (address[] memory) {
+    address[] memory watchList = s_watchList;
+    address[] memory needsFunding = new address[](watchList.length);
+    uint256 count = 0;
+    uint256 minWaitPeriod = s_minWaitPeriodSeconds;
+    uint256 contractBalance = ERC20Token.balanceOf(address(this));
+    Target memory target;
+    for (uint256 idx = 0; idx < watchList.length; idx++) {
+      target = s_targets[watchList[idx]];
+      uint256 targetTokenBalance = ERC20Token.balanceOf(watchList[idx]);
+      if (
+        target.lastTopUpTimestamp + minWaitPeriod <= block.timestamp &&
+        contractBalance >= target.topUpAmount &&
+        targetTokenBalance < target.minBalance
+      ) {
+        needsFunding[count] = watchList[idx];
+        count++;
+        contractBalance -= target.topUpAmount;
+      }
+    }
+    if (count != watchList.length) {
+      assembly {
+        mstore(needsFunding, count) // resize array to number of valid targets
+      }
+    }
+    return needsFunding;
+  }
+
+  /**
+   * @notice Send funds to the subscriptions provided.
+   * @param needsFunding the list of subscriptions to fund
+   */
+  function topUp(address[] memory needsFunding) public whenNotPaused {
+    uint256 minWaitPeriodSeconds = s_minWaitPeriodSeconds;
+    Target memory target;
+    for (uint256 idx = 0; idx < needsFunding.length; idx++) {
+      target = s_targets[needsFunding[idx]];
+      uint256 targetTokenBalance = ERC20Token.balanceOf(needsFunding[idx]);
+      uint256 contractBalance = ERC20Token.balanceOf(address(this));
+      if (
+        target.isActive &&
+        target.lastTopUpTimestamp + minWaitPeriodSeconds <= block.timestamp &&
+        targetTokenBalance < target.minBalance &&
+        contractBalance >= target.topUpAmount
+      ) {
+        bool success = ERC20Token.transfer(needsFunding[idx], target.topUpAmount);
```

Comment:
Saw your comment about this in the summary, I think the success value would be useful here, e.g. usdc can have blacklists for transfer which can cause this to fail